moodle/Sniffs/Access/DeprecatedCapabilitySniff.php

```diff
@@ -0,0 +1,111 @@
+
+use PHP_CodeSniffer\Files\File;
+use PHP_CodeSniffer\Sniffs\Sniff;
+
+class DeprecatedCapabilitySniff implements Sniff {
+
+    // phpcs:disable moodle.NamingConventions.ValidVariableName.MemberNameUnderscore
+    /**
+     * If we try to check this capability, a warning will be shown.
+     *
+     * @var array
+     */
+    public $capabilitiesWarningList = [];
+
+    /**
+     * If we try to check this capability, an error will be shown.
+     *
+     * @var array
+     */
+    public $capabilitiesErrorList = [];
+
+    // phpcs:enable
+
+    /**
+     * Access function type
+     *
+     * @var string[]
+     */
+    public $accessfunctions = ['has_access', 'has_capability'];
+
+    /**
+     * Register for open tag (only process once per file).
+     */
+    public function register() {
+        return array(T_OPEN_PARENTHESIS);
+    }
+
+    /**
+     * Processes php files and for required login checks if includeing config.php
+     *
+     * @param File $file The file being scanned.
+     * @param int $stackptr The position in the stack.
+     */
+    public function process(File $file, $stackptr) {
+        $functionnamepos = $file->findPrevious(T_STRING, $stackptr - 1);
+        if ($functionnamepos === false) {
+            return;
+        }
+        $tokens = $file->getTokens();
+        if (empty($tokens[$functionnamepos]) || !$this->is_an_access_function($tokens[$functionnamepos])) {
+            return;
+        }
+
+        $alldeprecated = array_merge($this->capabilitiesErrorList, $this->capabilitiesWarningList);
+        $closeparenthesispos = $file->findNext(T_CLOSE_PARENTHESIS, $stackptr + 1);
+        if ($closeparenthesispos === false) {
+            // Live coding, parse error or not a function call.
+            return;
+        }
+        $closeparenthesistoken = $tokens[$closeparenthesispos];
+        $starttoken = $closeparenthesistoken['parenthesis_opener'] + 1;
+        $endtoken = $closeparenthesistoken['parenthesis_closer'];
+        $values = $file->getTokensAsString($starttoken, $endtoken - $starttoken);
+        foreach ($alldeprecated as $capability) {
+            if (strpos($values, $capability) !== false) {
+                if (in_array($capability, $this->capabilitiesWarningList)) {
+                    $file->addWarning("The following capability '$capability' will be deprecated soon.", $starttoken,
+                        'DeprecatedCapability');
+                } else {
+                    $file->addError("The following capability '$capability' has been deprecated.", $starttoken,
+                        'DeprecatedCapability');
+                }
+            }
+        }
+    }
+
+    /**
+     * Is the current name an access function
+     *
+     * @param array $token current token
+     * @return bool true if the current methodname is access function.
+     */
+    protected function is_an_access_function(array $token) {
+        return in_array($token['content'], $this->accessfunctions);
```

Comment:
```suggestion
        return in_array($token['content'], $this->accessfunctions, true);
```

Reply:
Hi @jrchamp

Thanks for the quick feedback. I went back to the doc of in_array I have not used this parameter before, so good to know. I could probably have used isset also. I am not sure which one is quicker.
I will modify this once you will have a chance to look at the rest. The global approach needs to be validated and the big picture is: do we really get this the capabilitiesWarningList through rulseset.xml or by reading each access.php. 

Thanks, 

Laurent

Reply:
Hi @jrchamp,

Cheers, I will take that into account once the full review is done. As we say "Make it work, Make it better"...

To give you the context will have couple of capabilities in the array so (n) will be small and as per the second in_array, the number of parameters is also very small (2 or 3 max). I knew about the isset better performances, but then again if performance is not the main issue (and here we have only a couple of elements), I prefer in_array as it is more understandable as you read the code. 
I have never noticed with the third parameter of is_array thanks for the input and the detailed explanations !

Laurent